tkp/db/alchemy/image.py

```diff
@@ -0,0 +1,192 @@
+    and updates the assocskyrgn table accordingly.
+
+    Any previous entries in assocskyrgn relating to this skyregion are
+    deleted first.
+
+    Note 1. We use the variable 'inter' to cache the extraction_radius as transformed
+    onto the unit sphere, so this does not have to be recalculated for every
+    comparison.
+
+    Note 2. (To Do:) This distance check could be made more efficient by
+    restricting to a range of RA values, as we do with the Dec.
+    However, this optimization is complicated by the meridian wrap-around issue.
+        """
+    inter = 2. * math.sin(math.radians(skyregion.xtr_radius) / 2.)
+    inter_sq = inter * inter
+
+    q = """
+      INSERT INTO assocskyrgn
+        (
+        runcat
+        ,skyrgn
+        ,distance_deg
+        )
+      SELECT rc.id as runcat
+            ,sky.id as skyrgn
+            ,DEGREES(2 * ASIN(SQRT( (rc.x - sky.x) * (rc.x - sky.x)
+                                    + (rc.y - sky.y) * (rc.y - sky.y)
+                                    + (rc.z - sky.z) * (rc.z - sky.z)
+                                  ) / 2 )
+                     )
+        FROM skyregion sky
+            ,runningcatalog rc
+       WHERE sky.id = %(skyregion_id)s
+         AND rc.dataset = sky.dataset
+         AND rc.wm_decl BETWEEN sky.centre_decl - sky.xtr_radius
+                            AND sky.centre_decl + sky.xtr_radius
+         AND (  (rc.x - sky.x) * (rc.x - sky.x)
+                + (rc.y - sky.y) * (rc.y - sky.y)
+                + (rc.z - sky.z) * (rc.z - sky.z)
+             ) < %(inter_sq)s
+      ;
+    """ % {'inter_sq': inter_sq, 'skyregion_id': skyregion.id}
+    session.execute(q)
+    return inter
+
+
+def get_skyregion(session, dataset, centre_ra, centre_decl, xtr_radius):
+    """
+     gets an id for a skyregion, given a pair of central co-ordinates and a radius.
+
+     If no matching skyregion is found, a new one is inserted. In this case we also trigger execution of
+     `updateSkyRgnMembers` for the new skyregion - this performs a simple assocation with current members of the
+     runningcatalog to find sources that should be visible in the new skyregion, and updates the assocskyrgn table
+     accordingly.
+
+     args:
+        session (sqlalchemy.orm.session.Session): a SQLAlchemy session object
+        dataset_id (int): the dataset ID
+        centre_ra (float): center RA coordinate
+        centre_decl (float): center DECL coordinate
+        xtr_radius (float): The extraction radius
+
+     returns:
+        tkp.db.models.Skyregion: a SQLalchemy skyregion
+     """
+    skyregion = session.query(Skyregion).filter(Skyregion.dataset == dataset,
+                                                Skyregion.centre_ra == centre_ra,
+                                                Skyregion.centre_decl == centre_decl,
+                                                Skyregion.xtr_radius == xtr_radius).first()
```

Comment:
Using `.first` works fine here, but I would say [`.one_or_none`](http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.one_or_none) gives better documentation of intent - we don't expect there to ever be multiple identical Skyregions, so `first` looks a bit weird to me.